fs_impl.go

```diff
@@ -0,0 +1,353 @@
+	parent_record.Children = append(parent_record.Children, new_id)
+	fs.records[parent_uuid] = parent_record
+
+	return new_id, fs.writeToDisk(new_id, new_rec)
+}
+
+// return new slice that does not contain v
+func removeUUID(s []uuid.UUID, v uuid.UUID) []uuid.UUID {
+	i := 0
+	pos := -1
+	for ; i < len(s); i++ {
+		if s[i] == v {
+			pos = i
+			break
+		}
+	}
+
+	for ; i < len(s); i++ {
+		if s[i] == v {
+			panic("duplicite uuid")
+		}
+	}
+
+	if pos == -1 {
+		panic("uuid not found")
+	}
+
+	// swap remove
+	s[pos] = s[len(s)-1]
+	return s[:len(s)-1]
+}
+
+func (fs *Fs) deleteRecordFilesFromDisk(name string) error {
+	entries, err := os.ReadDir(fs.basePath)
+	if err != nil {
+		return err
+	}
+
+	// list of all errors. We try to delete as many files as we can to
+	// avoid zombie files and then return one big error with all of the
+	// failed removes.
+	var errs []error
+
+	err = os.Remove(fs.path(name))
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	// TODO(prokop) keep a list of all sections instead of iterating over
+	// all files
+	for _, e := range entries {
+		if strings.HasPrefix(e.Name(), name+".") {
+			err = os.Remove(fs.path(e.Name()))
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	if len(errs) != 0 {
+		return errors.Join(errs...)
+	}
+	return nil
+}
+
+func (fs *Fs) deleteChildInMemoryAndWriteParentToDisk(parent uuid.UUID, child uuid.UUID) error {
+	ch := fs.records[parent].Children
+	fs.records[parent].Children = removeUUID(ch, child)
+
+	return fs.writeToDisk(parent, fs.records[parent])
+}
+
+func (fs *Fs) decreseChildrenRefCountRecursive(current uuid.UUID) error {
+	// list of all errors. We try to delete as many files as we can to
+	// avoid zombie files and then return one big error with all of the
+	// failed removes.
+	var errs []error
+
+	for _, child := range fs.records[current].Children {
+		if fs.records[child].refs == 0 {
+			panic("decresing ref count below zero")
+		}
+
+		fs.records[child].refs--
+
+		if fs.records[child].refs == 0 {
+			err := fs.delete(current, child)
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	if len(errs) != 0 {
+		return errors.Join(errs...)
+	}
+
+	return nil
+}
+
+func checkSectionNameIsSane(section string) error {
+	match, _ := regexp.MatchString(onlySectionPattern, section)
+	if !match {
+		return errors.New("section name is not sane")
+	}
+	return nil
+}
+
+func (fs *Fs) getSectionFileName(file uuid.UUID, section string) string {
+	return fs.path(file.String() + "." + section)
+}
+
+// fileStorer impl
+
+func (fs *Fs) getChildren(u uuid.UUID) ([]uuid.UUID, error) {
+	return fs.records[u].Children, nil
+}
+
+func (fs *Fs) mkdir(parent_uuid uuid.UUID, name string) (uuid.UUID, error) {
+	return fs.newRecord(parent_uuid, name)
+}
+
+func (fs *Fs) touch(parent_uuid uuid.UUID, name string) (uuid.UUID, error) {
+	return fs.newRecord(parent_uuid, name)
+}
+
+func (fs *Fs) delete(parent uuid.UUID, child uuid.UUID) error {
```

Comment:
I beleive this is an incorrect behavior. This will delete the file, even if it's ref count may be non zero.

Reply:
my bad